crates/pipeline_manager/src/runner.rs

```diff
@@ -551,7 +552,500 @@ impl LocalRunner {
+            RunnerInner::Local(runner) => {
+                runner
+                    .notify_pipeline_automaton(tenant_id, pipeline_id)
+                    .await
+            }
+        }
+    }
+}
+
+/// A handle to the pipeline process that kills the pipeline
+/// on `drop`.
+struct PipelineHandle {
+    pipeline_process: Child,
+}
+
+impl Drop for PipelineHandle {
+    fn drop(&mut self) {
+        let _ = self.pipeline_process.kill();
+    }
+}
+
+/// Pipeline automaton monitors the runtime state of a single pipeline
+/// and continually reconciles desired and actual state.
+///
+/// The automaton runs as a separate tokio task.
+struct PipelineAutomaton {
+    pipeline_id: PipelineId,
+    tenant_id: TenantId,
+    pipeline_process: Option<PipelineHandle>,
+    config: Arc<CompilerConfig>,
+    db: Arc<Mutex<ProjectDB>>,
+    notifier: Arc<Notify>,
+}
+
+impl PipelineAutomaton {
+    /// The frequency of polling the pipeline during normal operation
+    /// when we don't normally expect its state to change.
+    const DEFAULT_PIPELINE_POLL_PERIOD: Duration = Duration::from_millis(10_000);
+
+    /// Max time to wait for the pipeline process to initialize its
+    /// HTTP server.
+    const PROVISIONING_TIMEOUT: Duration = Duration::from_millis(10_000);
+
+    /// How often to check for the pipeline port file during the
+    /// provisioning phase.
+    const PROVISIONING_POLL_PERIOD: Duration = Duration::from_millis(300);
+
+    /// Max time to wait for the pipeline to initialize all connectors.
+    const INITIALIZATION_TIMEOUT: Duration = Duration::from_millis(20_000);
+
+    /// How often to poll for the pipeline initialization status.
+    const INITIALIZATION_POLL_PERIOD: Duration = Duration::from_millis(300);
+
+    /// Max time to wait for the pipeline process to exit.
+    // TODO: It seems that Actix often takes a while to shutdown.  This
+    // is something to investigate.
+    const SHUTDOWN_TIMEOUT: Duration = Duration::from_millis(3_000);
+
+    /// How often to poll for the pipeline process to exit.
+    const SHUTDOWN_POLL_PERIOD: Duration = Duration::from_millis(300);
+
+    fn new(
+        pipeline_id: PipelineId,
+        tenant_id: TenantId,
+        config: &Arc<CompilerConfig>,
+        db: Arc<Mutex<ProjectDB>>,
+        notifier: Arc<Notify>,
+    ) -> Self {
+        Self {
+            pipeline_id,
+            tenant_id,
+            pipeline_process: None,
+            config: config.clone(),
+            db,
+            notifier,
+        }
+    }
+
+    /// Runs until the pipeline is deleted or an unexpected error occurs.
+    async fn run(self) -> Result<(), ManagerError> {
+        let pipeline_id = self.pipeline_id;
+
+        self.do_run().await.map_err(|e| {
+            error!(
+                "Pipeline automaton '{}' terminated with error: '{e}'",
+                pipeline_id
+            );
+            e
+        })
+    }
+
+    async fn do_run(mut self) -> Result<(), ManagerError> {
```

Comment:
This method is 300+ lines long. Much of the body could be refactored into smaller methods for readability.